src/reducers/platform.js

```diff
@@ -19,45 +20,52 @@ import {
 	CACHE_SUCCESS
 } from '../actions/cacheActionCreators';
 
+type ApiState = {
+	inFlight: Array<string>,
+	fail?: boolean,
+	[string]: QueryResponse,
+};
+
+export const DEFAULT_API_STATE: ApiState = { inFlight: [] };
 export const DEFAULT_APP_STATE = { isFetching: false };
 
-export const responseToState = response => {
-	const { ref, ...data } = response;
-	return { [ref]: data };
-};
+export const responseToState = (response: QueryResponse): { [string]: QueryResponse } =>
+	({ [response.ref]: response });
 
-/**
+/*
  * The primary reducer for data provided by the API
- * `state.api` sub-tree
- *
- * @param {Object} state
- * @param {ReduxAction} action
- * @return {Object}
  */
-export function api(state=DEFAULT_APP_STATE, action={}) {
+export function api(state: ApiState = DEFAULT_API_STATE, action: FluxStandardAction): ApiState {
 	switch (action.type) {
-	case API_REQ:
+	case API_REQ: {
+		const requestRefs = (action.payload || []).map(({ ref }) => ref);
+		const inFlight = state.inFlight
+			.filter(ref => !requestRefs.includes[ref]) // clean out current duplicates
+			.concat(requestRefs); // add new requested refs
+
 		if ((action.meta || {}).logout) {
 			// clear app state during logout
-			return { ...DEFAULT_APP_STATE, isFetching: true };
+			return { ...DEFAULT_API_STATE, inFlight };
 		}
-		return { ...state, isFetching: true };
+
+		return { ...state, inFlight };
+	}
 	case API_RESP_SUCCESS:  // fall though
-	case CACHE_SUCCESS:  // fall through
 	case API_RESP_ERROR:
+	case CACHE_SUCCESS:  // fall through
 		// each of these actions provides an API response that should go into app
 		// state - error responses will contain error info
 		delete state.fail;  // if there are any values, the API is not failing
-		return { ...state, ...responseToState(action.payload.response) };
+		return {
+			...state,
+			...responseToState((action.payload || {}).response),
+			inFlight: state.inFlight.filter(ref => ref !== (action.payload || {}).response.ref),
+		};
 	case API_RESP_FAIL:
-		state.fail = action.payload;
-		// fall through - fetch is complete
-	case API_RESP_COMPLETE:
-		return { ...state, isFetching: false };
-
-	default:
-		return state;
+		return { ...state, fail: action.payload };
```

Comment:
I'm a little unsure how to correctly handle `API_RESP_FAIL` and `API_RESP_COMPLETE`, because they currently don't report what refs they correspond to, so there's no clear way to accurately update `inFlight` when the response fails - the payload for `API_RESP_FAIL` is an `Error` object, and `API_RESP_COMPLETE` doesn't have a payload. Maybe those actions should have a `meta.queries` property so that it is easy to see what request(s) they correspond to. I'll make that update in this PR.